### PR TITLE
Part 4: add dialog and form components

### PR DESCRIPTION
**Pull request by K41S4**

### CODE REVIEW

File: src/Components/Dialog/Dialog.module.css [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  text-transform: uppercase;
+  font-family: 'Montserrat', sans-serif;
+}
+
+.modalHeader {
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  margin-bottom: 20px;
+}
```
Comment (sxurik) [inline src/Components/Dialog/Dialog.module.css]: `.closeButton` has `position: absolute;`, so `display: flex;` on `.modalHeader` is not actually needed.
`.closeButton` actually can be moved outside `.modalHeader`.
And there is no point in `.modalHeader` wrapper around dialog's title.

File: src/Components/Dialog/__tests__/Dialog.test.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+describe('Dialog', () => {
+  test('renders all details', () => {
+    const title = 'Title test';
+    const children = 'children test';
+
+    render(
+      <Dialog title={title} onClose={jest.fn()}>
+        <button>{children}</button>
+      </Dialog>
```
Comment (sxurik) [inline src/Components/Dialog/__tests__/Dialog.test.js]: `children` name for this variable is a little bit misleading.
For `Dialog` the whole `<button>{children}</button>` element is `children`.

File: src/Components/Dialog/__tests__/Dialog.test.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    );
+
+    expect(screen.getByText(title)).toBeInTheDocument();
+    expect(screen.getByRole('button', { name: children })).toBeInTheDocument();
+  });
+
+  test('handles close event with calling callback', async () => {
```
Comment (sxurik) [inline src/Components/Dialog/__tests__/Dialog.test.js]: Something like this will be more clear in my opinion:

```suggestion
  test('calls close callback on close button click', async () => {
```

or something like this.

File: src/Components/MovieForm/LabeledInput/LabeledInput.jsx [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import styles from './MovieForm.module.css';
```
Comment (sxurik) [inline src/Components/MovieForm/LabeledInput/LabeledInput.jsx]: Shouldn't it be:

```suggestion
import styles from './LabeledInput.module.css';
```

File: src/Components/MovieForm/MovieForm.jsx [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+export const MovieForm = ({ initialValues = {}, onSubmit, genres }) => {
+  const [movie, setMovie] = useState({
+    title: initialValues?.title ?? '',
+    releaseDate: initialValues?.releaseDate ?? '',
+    imageUrl: initialValues?.imageUrl ?? '',
+    rating: initialValues?.rating ?? '',
+    genre: initialValues?.genre ?? '',
+    runtime: initialValues?.runtime ?? '',
+    description: initialValues?.description ?? '',
```
Comment (sxurik) [inline src/Components/MovieForm/MovieForm.jsx]: Optional chaining (`?.`) is not needed here, because `initialValues` has default value of `{}`.

File: src/Components/MovieForm/MovieForm.jsx [hunk 6]
```diff
@@ -0,0 +1,100 @@
+import { useState } from 'react';
+import styles from './MovieForm.module.css';
+import { LabeledInput } from './LabeledInput/LabeledInput';
+
+export const MovieForm = ({ initialValues = {}, onSubmit, genres }) => {
```
Comment (sxurik) [inline src/Components/MovieForm/MovieForm.jsx]: You can actually define default value for `initialValues` not as an empty object, but as object with empty strings, that can be used here and in `handleReset`.

File: src/Components/MovieForm/MovieForm.jsx [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      className={styles.container}
+      onSubmit={(e) => {
+        e.preventDefault();
+        const formData = Object.fromEntries(new FormData(e.target));
+        onSubmit(formData);
+      }}
+    >
+      <div className={styles.formRow}>
+        <div className={styles.formGroup}>
+          <LabeledInput label={'Title'} onChange={handleChange} value={movie.title} id={'title'} />
```
Comment (sxurik) [inline src/Components/MovieForm/MovieForm.jsx]:
```suggestion
          <LabeledInput label='Title' onChange={handleChange} value={movie.title} id='title' />
```

And in other places like this.

File: src/Components/MovieForm/MovieForm.jsx [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+          Description
+        </label>
+        <textarea
+          id="description"
+          name="description"
+          className={styles.overviewTextarea}
+          rows="4"
+          value={movie.description}
+          onChange={handleChange}
+        ></textarea>
```
Comment (sxurik) [inline src/Components/MovieForm/MovieForm.jsx]: `textarea` is a self-closing element, so something lke this is more common:


```suggestion
        <textarea
          id="description"
          name="description"
          className={styles.overviewTextarea}
          rows="4"
          value={movie.description}
          onChange={handleChange}
        />
```

File: src/Components/MovieForm/MovieForm.jsx [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+      </div>
+
+      <div className={styles.formActions}>
+        <button className={styles.resetButton} type="reset" onClick={handleReset}>
+          Reset
+        </button>
+        <button className={styles.submitButton}>Submit</button>
```
Comment (sxurik) [inline src/Components/MovieForm/MovieForm.jsx]: To be more explicit:

```suggestion
        <button type="submit" className={styles.submitButton}>Submit</button>
```

File: src/Components/MovieForm/MovieForm.jsx [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+      </div>
+
+      <div className={styles.formGroup}>
+        <label htmlFor="description" className={styles.overviewLabel}>
+          Description
+        </label>
+        <textarea
+          id="description"
+          name="description"
+          className={styles.overviewTextarea}
```
Comment (sxurik) [inline src/Components/MovieForm/MovieForm.jsx:82]: Max width is usually set for resizable textareas to avoid this:

image

File: src/Components/MovieForm/MovieForm.jsx [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+          <LabeledInput label={'Rating'} onChange={handleChange} value={movie.rating} id={'rating'} />
+        </div>
+      </div>
+      <div className={styles.formRow}>
+        <div className={styles.formGroup}>
+          <label className={styles.label} htmlFor={'genres'}>
+            Genre
+          </label>
+          <select className={styles.select} id={'genres'} name="genre" defaultValue={movie.genre}>
+            <option hidden></option>
```
Comment (sxurik) [inline src/Components/MovieForm/MovieForm.jsx]: Self-closing as well:

```suggestion
            <option hidden />
```

Comment (sxurik) [reply src/Components/MovieForm/MovieForm.jsx]: Maybe it makes sense to add `prettier` to the project so that it formats everything consistently.

File: src/Components/MovieForm/__tests__/MovieForm.test.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    expect(titleInput).toHaveValue(movieDetails.title);
+    expect(releaseDateInput).toHaveValue(movieDetails.releaseDate);
+    expect(imageUrlInput).toHaveValue(movieDetails.imageUrl);
+    expect(ratingInput).toHaveValue(movieDetails.rating);
+    expect(genreInput).toHaveValue(movieDetails.genre);
+    expect(runtimeInput).toHaveValue(movieDetails.runtime);
+    expect(descriptionTextarea).toHaveValue(movieDetails.description);
+  });
+
+  it('handles submit event with correct values', async () => {
```
Comment (sxurik) [inline src/Components/MovieForm/__tests__/MovieForm.test.js]:
```suggestion
  it('handles submit event with default values', async () => {
```

File: src/Components/MovieForm/__tests__/MovieForm.test.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    await user.type(runtimeInput, movieDetails.runtime);
+    await user.type(descriptionTextarea, movieDetails.description);
+
+    expect(titleInput).toHaveValue(movieDetails.title);
+    expect(releaseDateInput).toHaveValue(movieDetails.releaseDate);
+    expect(imageUrlInput).toHaveValue(movieDetails.imageUrl);
+    expect(ratingInput).toHaveValue(movieDetails.rating);
+    expect(genreInput).toHaveValue(movieDetails.genre);
+    expect(runtimeInput).toHaveValue(movieDetails.runtime);
+    expect(descriptionTextarea).toHaveValue(movieDetails.description);
```
Comment (sxurik) [inline src/Components/MovieForm/__tests__/MovieForm.test.js]: Instead of these many `expect`s you can use the same approach as in your `it('handles submit event with correct values'` test. Just check that `mockSubmit` is called with new values you use in ` user.type(`.

Or you can use `it.each` for all `text` inputs.

File: src/Components/MovieForm/__tests__/MovieForm.test.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+      rating: '9',
+      genre: 'Action',
+      runtime: '1h',
+      description: 'Test description',
+    };
+
+    const { asFragment } = render(<MovieForm initialValues={movieDetails} onSubmit={jest.fn()} genres={genres} />);
+
+    expect(asFragment()).toMatchSnapshot();
+  });
```
Comment (sxurik) [inline src/Components/MovieForm/__tests__/MovieForm.test.js]: You don't actually need snapshot test here. You should check `initialValues` using `expect(mockSubmit).toHaveBeenCalledWith(...)`

File: src/__mocks__/tabbable.js
Comment (sxurik) [inline src/__mocks__/tabbable.js:1]: Why is this needed?

File: src/stories/MovieFormComposition/Composition.module.css [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+  margin-bottom: 5px;
+  color: #ffffff;
+}
+
+button {
+  background-color: #f65261;
+  color: #ffffff;
+  border: none;
+  padding: 10px 40px;
+}
```
Comment (sxurik) [inline src/stories/MovieFormComposition/Composition.module.css]: Please, use class names.
Because you used tag for styling here, these styles were applied to the "Close" button, not only to "Confirm":

image

File: src/stories/Dialog/Dialog.stories.jsx [hunk 16]
```diff
@@ -0,0 +1,25 @@
+import { fn } from '@storybook/test';
+
+import { Dialog } from '../../Components/Dialog/Dialog';
+
+export default {
+  title: 'Example/Dialog',
```
Comment (sxurik) [inline src/stories/Dialog/Dialog.stories.jsx]: For all `Dialog` stories could you please render a button, that displays the `Dialog` on clickand add `onClose` logic?
Otherwise there is no way to check documentation of the component, the Dialog renders on story open and covers everything:

image

File: src/stories/MovieFormComposition/Composition.module.css [hunk 17]
```diff
@@ -0,0 +1,11 @@
+.confirmationText {
```
Comment (sxurik) [inline src/stories/MovieFormComposition/Composition.module.css]: Maybe this file should be called `MovieFormComposition.module.css` or `DeleteMovie.stories.module.css`?

File: src/__mocks__/tabbable.js
Comment (K41S4) [reply src/__mocks__/tabbable.js:1]: i get this error without the mock
image

File: src/Components/MovieForm/MovieForm.jsx [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+export const MovieForm = ({ initialValues = defaultMovieValues, onSubmit, genres }) => {
+  const [movie, setMovie] = useState({
+    title: initialValues.title ?? '',
+    releaseDate: initialValues.releaseDate ?? '',
+    imageUrl: initialValues.imageUrl ?? '',
+    rating: initialValues.rating ?? '',
+    genre: initialValues.genre ?? '',
+    runtime: initialValues.runtime ?? '',
+    description: initialValues.description ?? '',
+  });
```
Comment (sxurik) [inline src/Components/MovieForm/MovieForm.jsx]: Can it be simplified it with something like this?

```suggestion
  const [movie, setMovie] = useState({
    ...defaultMovieValues,
    ...initialValues,
  });
```

**Review by sxurik — APPROVED** [review]
